tools/amd_build/pyHIPIFY/hipify-python.py

```diff
@@ -0,0 +1,966 @@
+                # Updates kernel
+                kernel_with_template = "%s<real>" % (kernel_name)
+            else:
+                kernel_with_template = kernel_name
+            formatted_args = {}
+            for idx, arg_type in enumerate(argument_types):
+                formatted_args[idx] = arg_type
+
+            KernelDictionary[kernel_name] = {"kernel_with_template": kernel_with_template, "arg_types": formatted_args}
+
+        # Extract generated kernels
+        get_generated_kernels = [k for k in re.finditer(r"GENERATE_KERNEL([1-9])\((.*)\)", string)]
+        # curandStateMtgp32 *state, int size, T *result, ARG1
+        for kernel in get_generated_kernels:
+            kernel_gen_type = int(kernel.group(1))
+            kernel_name = kernel.group(2).split(",")[0]
+            kernel_params = kernel.group(2).split(",")[1:]
+
+            if kernel_gen_type == 1:
+                kernel_args = {1: "int", 2: "%s *" % kernel_params[0], 3: kernel_params[1]}
+
+            if kernel_gen_type == 2:
+                kernel_args = {1: "int", 2: "%s *" % kernel_params[0], 3: kernel_params[1], 4: kernel_params[2]}
+
+            # Argument at position 1 should be int
+            KernelDictionary[kernel_name] = {"kernel_with_template": kernel_name, "arg_types": kernel_args}
+
+
+def disable_unsupported_function_call(function, input_string, replacement):
+    """Disables calls to an unsupported HIP function"""
+    # Prepare output string
+    output_string = input_string
+
+    # Find all calls to the function
+    calls = re.finditer(r"\b%s\b" % function, input_string)
+
+    # Do replacements
+    for call in calls:
+        start = call.start()
+        end = call.end()
+
+        pos = end
+        started_arguments = False
+        bracket_count = 0
+        while pos < len(input_string):
+            if input_string[pos] == "(":
+                if started_arguments is False:
+                    started_arguments = True
+                    bracket_count = 1
+                else:
+                    bracket_count += 1
+            elif input_string[pos] == ")" and started_arguments:
+                bracket_count -= 1
+
+            if bracket_count == 0 and started_arguments:
+                # Finished!
+                break
+            pos += 1
+
+        function_call = input_string[start:pos + 1]
+        output_string = output_string.replace(function_call, replacement)
+
+    return output_string
+
+
+def disable_module(input_file):
+    """Disable a module entirely except for header includes."""
+    with openf(input_file, "r+") as f:
+        txt = f.read()
+        last = list(re.finditer(r"#include .*\n", txt))[-1]
+        end = last.end()
+
+        disabled = "%s#if !defined(__HIP_PLATFORM_HCC__)\n%s\n#endif" % (txt[0:end], txt[end:])
+
+        f.seek(0)
+        f.write(disabled)
+        f.truncate()
+
+
+def extract_arguments(start, string):
```

Comment:
So, IIUC, this requires you to pass it a string that starts with a parentheses, right? That's a very important invariant that should be documented here.

Reply:
The string doesn't have to *start* with a parenthesis, but does need to have the arguments enclosed within parentheses as part of the string

Reply:
Function requirement clarified in comments with an example in commit acfacbc as part of PR https://github.com/ROCmSoftwarePlatform/pytorch/pull/20